File: experiments/gt-p139-report-data/footprint.py
#!/usr/bin/env python3
"""P139: distinct instruction bytes one KEM operation executes, from a callgrind profile collected
only inside that entry point (--collect-atstart=no --toggle-collect=crypto_kem_<op>
--dump-instr=yes --dump-line=no --compress-pos=no --compress-strings=no).  AArch64 instructions
are 4 bytes.  Library code (libc's memcpy/memset/... linked statically) is counted apart.
(lackey's --trace-mem was the first choice; on the Pi it spins, so callgrind it is.)
usage: footprint.py CALLGRIND_OUT"""
import sys, re
LIB = re.compile(r'^(_?_?mem|_?_?bzero|explicit_bzero|__explicit|__libc|_dl|__GI|__stack_chk|strlen|__strlen)')
fn = None; own = set(); lib = set()
for line in open(sys.argv[1]):
    if line.startswith('fn='): fn = line[3:].strip(); continue
    if not line.startswith('0x'): continue
    p = line.split(); a = int(p[0], 16)
    (lib if LIB.match(fn or '') else own).add(a)
print(f'{4 * len(own)} bytes KEM code, {4 * len(lib)} bytes library code')

File: experiments/gt-p139-report-data/footprint.c
/* P139: code executed by one KEM operation, measured by callgrind collecting only inside the entry
 * point (run_extra.sh, footprint.py).  The set-up key pair and encapsulation take the same path as the
 * measured call, so they add no distinct instructions.  usage: footprint 0|1|2 */
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include "api.h"
void rb_seed(void);
__attribute__((noinline)) void marker_begin(void){ __asm__ volatile("" ::: "memory"); }
__attribute__((noinline)) void marker_end(void){ __asm__ volatile("" ::: "memory"); }
static uint8_t pk[CRYPTO_PUBLICKEYBYTES], sk[CRYPTO_SECRETKEYBYTES], ct[CRYPTO_CIPHERTEXTBYTES], ss[CRYPTO_BYTES];
int main(int argc, char **argv){
  int op = argc > 1 ? atoi(argv[1]) : 0;
  rb_seed(); crypto_kem_keypair(pk, sk); crypto_kem_enc(ct, ss, pk);
  rb_seed();
  marker_begin();
  if (op == 0) crypto_kem_keypair(pk, sk); else if (op == 1) crypto_kem_enc(ct, ss, pk); else crypto_kem_dec(ss, ct, sk);
  marker_end();
  return 0; }
